prbench/src/prbench/envs/geom3d/packing3d.py

```diff
@@ -0,0 +1,765 @@
+                self._part_ids_to_triangle_features[part_id] = (
+                    side_a,
+                    side_b,
+                    depth,
+                    triangle_type,
+                )
+
+                # Place part on table while avoiding collisions with other parts and
+                # the rack (we allow parts to start outside the rack)
+                for _ in range(100_000):
+                    # Sample a pose on the table surface.
+                    x = self.np_random.uniform(
+                        self.config.table_pose.position[0]
+                        - self.config.table_half_extents[0]
+                        + half_extents[0],
+                        self.config.table_pose.position[0]
+                        + self.config.table_half_extents[0] * 0.5
+                        - half_extents[0],
+                    )
+                    y = self.np_random.uniform(
+                        self.config.table_pose.position[1]
+                        - self.config.table_half_extents[1]
+                        + half_extents[1],
+                        self.config.table_pose.position[1]
+                        + self.config.table_half_extents[1]
+                        - half_extents[1],
+                    )
+                    z = (
+                        self.config.table_pose.position[2]
+                        + self.config.table_half_extents[2]
+                        + part_z_half_extent
+                    )
+                    set_pose(part_id, Pose((x, y, z)), self.physics_client_id)
+
+                    collision_exists = False
+                    for other_id in ({self._rack_id} | set(self._part_ids.values())) - {
+                        part_id
+                    }:
+                        if check_body_collisions(
+                            part_id, other_id, self.physics_client_id
+                        ):
+                            collision_exists = True
+                            break
+
+                        # distance = np.linalg.norm(
+                        #     np.array(p.getBasePositionAndOrientation(part_id)[0])
+                        #     - np.array(p.getBasePositionAndOrientation(other_id)[0])
+                        # )
+                        # if distance < 0.2:  # minimum distance between parts
+                        #     collision_exists = True
+                        #     break
+
+                    if not collision_exists:
+                        break
+                else:
+                    raise RuntimeError("Failed to sample part pose")
+
+            else:
+                raise ValueError(f"Unsupported part type: {part_type}")
+
+    def _set_object_states(self, obs: Geom3DObjectCentricState) -> None:
+        assert isinstance(obs, Packing3DObjectCentricState)
+        # Update rack (recreate if half extents changed)
+        if self._rack_half_extents != getattr(
```

Comment:
why is getattr necessary here and below as opposed to just `obs.rack_half_extents`?